leases: add more details on API Server Identity (#38196)
kubernetes/website: content/en/docs/concepts/architecture/leases.md

HA configurations, where only one instance of the component should be actively running while the other
instances are on stand-by.

> ## API Server Identity

Reviewer: ```suggestion
## API server identity
```
or
```suggestion
## kube-apiserver identity {#api-server-identity}
```
(essentially: retain the original). Only kube-apiserver implements this mechanism as far as I know.

Reviewer 2: All API servers with storage are expected to implement this (i.e. aggregated API servers and Kube API server).  This is currently broken outside of KAS though.

Reviewer 2: > OK. Presumably the expectation only comes in once the feature is GA?

This was something we missed by accident.  It should have been part of the beta.

discover how many instances of `kube-apiserver` are operating the Kubernetes control plane.
Existence of kube-apiserver leases enables future capabilities that may require coordination between
each kube-apiserver.

You can inspect Leases owned by each kube-apiserver by checking for lease objects in the `kube-system` namespace
with the name `kube-apiserver-<sha256-hash>`. Alternatively you can use the label selector `k8s.io/component=kube-apiserver`:

```
$ kubectl -n kube-system get lease -l k8s.io/component=kube-apiserver
NAME                                        HOLDER                                                                           AGE

> kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a   kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a_9cbf54e5-1136-44bd-8f9a-1dcd15c346b4   5m33s

Reviewer 2: Should we include the YAML of one of these as an example?  To show all the labels?

`kubectl -n kube-system get lease kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a -o yaml`?

Author: done

Reviewer: See https://github.com/kubernetes/website/pull/38196#discussion_r1037486056

kube-apiserver-dz2dqprdpsgnm756t5rnov7yka   kube-apiserver-dz2dqprdpsgnm756t5rnov7yka_84f2a85d-37c1-4b14-b6b9-603e62e4896f   4m23s
kube-apiserver-fyloo45sdenffw2ugwaz3likua   kube-apiserver-fyloo45sdenffw2ugwaz3likua_c5ffa286-8a9a-45d4-91e7-61118ed58d2e   4m43s
```

The SHA256 hash used in the lease name is based on the OS hostname as seen by kube-apiserver. Each kube-apiserver should be

> configured to use a hostname that is unique within the cluster. New instances of kube-apiserver that use the same hostname

Reviewer 2: Can we document the use of the `kubernetes.io/hostname` label as well?

Reviewer: See https://github.com/kubernetes/website/pull/38196#discussion_r1037486056

```
$ kubectl -n kube-system get lease -l k8s.io/component=kube-apiserver
NAME                                        HOLDER                                                                           AGE
kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a   kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a_9cbf54e5-1136-44bd-8f9a-1dcd15c346b4   5m33s
kube-apiserver-dz2dqprdpsgnm756t5rnov7yka   kube-apiserver-dz2dqprdpsgnm756t5rnov7yka_84f2a85d-37c1-4b14-b6b9-603e62e4896f   4m23s
kube-apiserver-fyloo45sdenffw2ugwaz3likua   kube-apiserver-fyloo45sdenffw2ugwaz3likua_c5ffa286-8a9a-45d4-91e7-61118ed58d2e   4m43s
```

The SHA256 hash used in the lease name is based on the OS hostname as seen by kube-apiserver. Each kube-apiserver should be
configured to use a hostname that is unique within the cluster. New instances of kube-apiserver that use the same hostname
will take over existing Leases using a new holder identity, as opposed to instantiating new lease objects. You can check the
hostname used by kube-apisever by checking the value of the `kubernetes.io/hostname` label:

```
$ kubectl -n kube-system get lease kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a -o yaml

> apiVersion: coordination.k8s.io/v1

Reviewer 2: Can you split this into two code blocks, one with `shell` syntax, the other with `yaml` syntax?

Author: done!

> discover how many instances of `kube-apiserver` are operating the Kubernetes control plane.
> Existence of kube-apiserver leases enables future capabilities that may require coordination between
> each kube-apiserver.

Reviewer: ```suggestion
The API server creates these leases in the `kube-system` namespace.
The existence of kube-apiserver Leases also enables adding future capabilities to Kubernetes,
that may require coordination between each kube-apiserver (for example, during cluster
upgrades).
```

> You can inspect Leases owned by each kube-apiserver by checking for lease objects in the `kube-system` namespace

Reviewer: ```suggestion
You can inspect Leases owned by each kube-apiserver by checking for Leases in the `kube-system` namespace
```

with the name `kube-apiserver-<sha256-hash>`. Alternatively you can use the label selector `k8s.io/component=kube-apiserver`:

```shell
$ kubectl -n kube-system get lease -l k8s.io/component=kube-apiserver

> NAME                                        HOLDER                                                                           AGE

Reviewer: Please split this - see [don't include the command prompt](https://kubernetes.io/docs/contribute/style/style-guide/#don-t-include-the-command-prompt) and [separate commands from output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output).

kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a   kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a_9cbf54e5-1136-44bd-8f9a-1dcd15c346b4   5m33s
kube-apiserver-dz2dqprdpsgnm756t5rnov7yka   kube-apiserver-dz2dqprdpsgnm756t5rnov7yka_84f2a85d-37c1-4b14-b6b9-603e62e4896f   4m23s
kube-apiserver-fyloo45sdenffw2ugwaz3likua   kube-apiserver-fyloo45sdenffw2ugwaz3likua_c5ffa286-8a9a-45d4-91e7-61118ed58d2e   4m43s
```

> The SHA256 hash used in the lease name is based on the OS hostname as seen by kube-apiserver. Each kube-apiserver should be

Reviewer: Which is true:
```suggestion
The SHA256 hash used in the Lease name is based on the OS hostname as seen by kube-apiserver. 
The API server coerces its hostname string to lowercase before hashing it. Each kube-apiserver
should be
```
or
```suggestion
The SHA256 hash used in the Lease name is based on the OS hostname as seen by kube-apiserver. 
Each kube-apiserver should be
```
?

configured to use a hostname that is unique within the cluster. New instances of kube-apiserver that use the same hostname

> will take over existing Leases using a new holder identity, as opposed to instantiating new lease objects. You can check the

Reviewer: ```suggestion
will take over existing Leases using a new holder identity, as opposed to instantiating new Leases.
You can check the
```

hostname used by kube-apisever by checking the value of the `kubernetes.io/hostname` label:

```shell

> $ kubectl -n kube-system get lease kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a -o yaml

Reviewer: (nit)
```suggestion
kubectl -n kube-system get lease kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a -o yaml
```
Maybe use a JSON path to get just the one label?

```

```yaml
apiVersion: coordination.k8s.io/v1
kind: Lease
metadata:

>   creationTimestamp: "2022-11-30T15:37:15Z"

Reviewer: (nit) Ideally, pick a date that is close to the v1.26 release.

  labels:
    k8s.io/component: kube-apiserver

>     kubernetes.io/hostname: kind-control-plane

Reviewer: https://kubernetes.io/docs/reference/labels-annotations-taints/ lists that this label is used on Nodes (only)

Either:
- fix the Kubernetes code not to label Leases with this official label
  - allowed, but a bit weird to make up a new one that is so similar
- update https://kubernetes.io/docs/reference/labels-annotations-taints/ to record that it is also used for leases that are associated with nodes

  name: kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a
  namespace: kube-system
  resourceVersion: "18171"
  uid: d6c68901-4ec5-4385-b1ef-2d783738da6c
spec:
  holderIdentity: kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a_9cbf54e5-1136-44bd-8f9a-1dcd15c346b4
  leaseDurationSeconds: 3600

>   renewTime: "2022-11-30T18:04:27.912073Z"

Reviewer: (nit) Ideally, pick a date that is close to the v1.26 release.

```

> Expired leases from kube-apiservers that no longer exist are garbage collected by new kube-apiservers after 1 hour.

Reviewer: Is this right:
```suggestion
### API server Lease garbage collection

Expired Leases from kube-apiservers that no longer exist are garbage collected by live kube-apiservers.
The control plane leaves the expired Leases for at least one hour before removing them.
```
?

We should update https://kubernetes.io/docs/concepts/architecture/garbage-collection/ to link to that heading, which is why I added it.
